osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -109,28 +110,38 @@ public partial class BeatmapCarousel : CompositeDrawable, IKeyBindingHandler<Glo
         [Cached]
         protected readonly CarouselScrollContainer Scroll;
 
+        [Resolved]
+        private RealmAccess realm { get; set; } = null!;
+
+        [Resolved]
+        private DetachedBeatmapStore? detachedBeatmapStore { get; set; }
+
+        private IBindableList<BeatmapSetInfo>? detachedBeatmapSets;
+
         private readonly NoResultsPlaceholder noResultsPlaceholder;
 
         private IEnumerable<CarouselBeatmapSet> beatmapSets => root.Items.OfType<CarouselBeatmapSet>();
 
-        // todo: only used for testing, maybe remove.
-        private bool loadedTestBeatmaps;
-
         public IEnumerable<BeatmapSetInfo> BeatmapSets
         {
             get => beatmapSets.Select(g => g.BeatmapSet);
             set
             {
-                loadedTestBeatmaps = true;
-                Schedule(() => loadBeatmapSets(value));
+                if (LoadState != LoadState.NotLoaded)
+                    throw new InvalidOperationException("If not using a realm source, beatmap sets must be set before load.");
```

Comment:
If this sort of thing is gonna get thrown then maybe this entire property should be `internal` to begin with?

Reply:
Yeah fair. It's only used for testing. `internal` doesn't really protect us against much, but it's a visual hint I guess.